lightningd/invoice.c

```diff
@@ -1219,29 +1223,31 @@ static void json_add_invoices(struct json_stream *response,
 			      const struct sha256 *payment_hash,
 			      const struct sha256 *local_offer_id)
 {
-	struct invoice_iterator it;
 	const struct invoice_details *details;
-	struct invoice invoice;
+	u64 inv_dbid;
 
 	/* Don't iterate entire db if we're just after one. */
 	if (label) {
-		if (wallet_invoice_find_by_label(wallet, &invoice, label)) {
+		if (invoices_find_by_label(wallet->invoices, &inv_dbid, label)) {
 			details =
-			    wallet_invoice_details(tmpctx, wallet, invoice);
+			    invoices_get_details(tmpctx, wallet->invoices, inv_dbid);
 			json_add_invoice(response, NULL, details);
 		}
 	} else if (payment_hash != NULL) {
-		if (wallet_invoice_find_by_rhash(wallet, &invoice,
+		if (invoices_find_by_rhash(wallet->invoices, &inv_dbid,
 						 payment_hash)) {
 			details =
-			    wallet_invoice_details(tmpctx, wallet, invoice);
+			    invoices_get_details(tmpctx, wallet->invoices, inv_dbid);
 			json_add_invoice(response, NULL, details);
 		}
 	} else {
-		memset(&it, 0, sizeof(it));
-		while (wallet_invoice_iterate(wallet, &it)) {
-			details = wallet_invoice_iterator_deref(response,
-								wallet, &it);
+		struct db_stmt *stmt;
+
+		for (stmt = invoices_first(wallet->invoices, &inv_dbid);
+		     stmt;
+		     stmt = invoices_next(wallet->invoices, stmt, &inv_dbid)) {
+			details = invoices_get_details(tmpctx,
+						       wallet->invoices, inv_dbid);
```

Comment:
This results in quite a lot more queries to populate listinvoices than it did previously.  This seems like a safer way to implement it, but there may be a performance impact for some nodes. I'm not sure how to evaluate the impact or if it's worth an optimization.

Reply:
Yep, looks like the dreaded n+1 issue (1 list query + n details/dependency queries). It might be more sensible to move the details into the main query, and iterate over the fully materialized details rather than the summary.

Then again, I expect this to be mostly left-over from the previous state.

Reply:
Yes, AFAICT this is no worse than the old code?

But a direct iterator would be more optimal.